packages/back-end/src/services/experimentQueries/experimentQueries.ts

```diff
@@ -0,0 +1,305 @@
+  factMetricGroups: FactMetricInterface[][];
+  // Legacy metrics always as singletons
+  legacyMetricSingles: MetricInterface[];
+}
+
+export function getFactMetricGroups(
+  metrics: ExperimentMetricInterface[],
+  settings: ExperimentSnapshotSettings,
+  integration: SourceIntegrationInterface,
+  organization: OrganizationInterface,
+): GroupedMetrics {
+  const legacyMetrics: MetricInterface[] = metrics.filter((m) =>
+    isLegacyMetric(m),
+  );
+  const factMetrics: FactMetricInterface[] = metrics.filter(isFactMetric);
+
+  const defaultReturn: GroupedMetrics = {
+    // by default, put all fact metrics in their own group
+    factMetricGroups: factMetrics.map((m) => [m]),
+    legacyMetricSingles: legacyMetrics,
+  };
+
+  // Combining metrics in a single query is an Enterprise-only feature
+  if (!orgHasPremiumFeature(organization, "multi-metric-queries")) {
+    return defaultReturn;
+  }
+
+  // Metrics might have different conversion windows which makes the query complicated
+  // TODO(sql): join together metrics with the same date windows for some added efficiency
+  if (settings.skipPartialData) {
+    return defaultReturn;
+  }
+
+  // Org-level setting (in case the multi-metric query introduces bugs)
+  // TODO(sql): deprecate this setting and hide it for orgs that have not set it
+  if (organization.settings?.disableMultiMetricQueries) {
+    return defaultReturn;
+  }
+
+  // Group fact metrics into efficient groups (primarily if they share a fact table)
+  const groups: Record<string, FactMetricInterface[]> = {};
+  factMetrics.forEach((m) => {
+    // Skip grouping metrics with percentile caps or quantile metrics if there's not an efficient implementation
+    if (
+      (m.cappingSettings.type === "percentile" || quantileMetricType(m)) &&
+      !integration.getSourceProperties().hasEfficientPercentiles
+    ) {
+      return;
+    }
+
+    const group = getFactMetricGroup(m);
+    if (group) {
+      groups[group] = groups[group] || [];
+      groups[group].push(m);
+    }
+  });
+
+  const groupArrays: FactMetricInterface[][] = [];
+  Object.values(groups).forEach((group) => {
+    // Split groups into chunks of MAX_METRICS_PER_QUERY
+    const chunks = chunkMetrics({
+      metrics: group.map((m) => {
+        const metric = cloneDeep(m);
+        // TODO(overrides): refactor overrides to beginning of analysis
+        applyMetricOverrides(metric, settings);
+        return {
+          metric,
+          regressionAdjusted:
+            isRegressionAdjusted(metric) &&
+            settings.regressionAdjustmentEnabled,
+        };
+      }),
+      maxColumnsPerQuery: integration.getSourceProperties().maxColumns,
+      bandit: !!settings.banditSettings,
+    });
+    groupArrays.push(...chunks);
+  });
+
+  // Add unused fact metrics as singles to the group array
+  factMetrics.forEach((m) => {
+    if (!groupArrays.some((group) => group.some((g) => g.id === m.id))) {
+      groupArrays.push([m]);
+    }
+  });
```

Comment:
This is a nested iteration (looking over all items in all the groups for each factMetric).

I think it would be fine given the low number of metrics/groups, but we can also do:
```
const groupedMetricIds = new Set(
	groupArrays.flatMap((group) => group.map((g) => g.id))
);
factMetrics.forEach((m) => {
	if (!groupedMetricIds.has(m.id)) { 
		groupArrays.push([m]);
	}
});
```

Reply:
Yeah I was just copying the original logic, I'll use your logic, it's cleaner.